server/schedule/operator/builder.go

```diff
@@ -449,33 +473,26 @@ func (b *Builder) planAddPeer() stepPlan {
 		a := b.toAdd.Get(i)
 		for _, leader := range b.currentPeers.IDs() {
 			if b.allowLeader(b.currentPeers.Get(leader)) {
-				best = b.comparePlan(best, stepPlan{addOrPromote: a, leader1: leader})
+				best = b.comparePlan(best, stepPlan{add: a, leaderAdd: leader})
 			}
 		}
 	}
 	return best
 }
 
-func (b *Builder) planPromotePeer() stepPlan {
-	for _, i := range b.toPromote.IDs() {
-		p := b.toPromote.Get(i)
-		return stepPlan{addOrPromote: p, leader1: b.currentLeader}
-	}
-	return stepPlan{}
-}
-
 // Pick the better plan from 2 candidates.
 func (b *Builder) comparePlan(best, next stepPlan) stepPlan {
 	if best.empty() {
 		return next
 	}
 	fs := []func(stepPlan) int{
+		b.preferReplaceByNearest,
 		b.avoidRemoveLeader,
 		b.preferUpStoreAsLeader,
 		b.preferOldPeerAsLeader,
-		b.preferLessLeaderTransfer1,
-		b.preferLessLeaderTransfer2,
+		b.preferLessLeaderTransfer,
 		b.preferTargetLeader,
+		b.preferAddOrPromoteTargetLeader,
```

Comment:
Why put `preferReplaceByNearest` to the first? I think it is less important:
+ `avoidRemoveLeader` is the most important one because breaking it may cause the operator be refused.
+ `preferOldPeerAsLeader` is also important as the issue #3819, but I think it will cause the operator failed only when the leader is transferred to the most recently added peer, which makes it not as important as `avoidRemoveLeader`.

IMO, `preferReplaceByNearest` is unlikely to cause operator fail, so it is at least less important than the above two.

What's more, I also think the order of the last three should be reversed.
+ "Add or promote target leader" is the precondition of "transfer to target leader". (`AddOrPromoteTargetLeader` > `TargetLeader`)
+ `preferLessLeaderTransfer` is just a normal optimization. Instead, `preferAddOrPromoteTargetLeader` is an optimization that try to avoid issue #3819. (`AddOrPromoteTargetLeader` > `LessLeaderTransfer`)
+ "Add or promote target leader" and "transfer to target leader" may also reduce "leader transfer" in the later steps. (`TargetLeader` > `LessLeaderTransfer`)

Reply:
I think `preferReplaceByNearest` should go before `preferOldPeerAsLeader` as ``preferOldPeerAsLeader` won't fail to operator, only make it slow.
I'm going to make it this way:
```
b.avoidRemoveLeader,      // 1. violate it causes the operator fail to execute.
b.preferReplaceByNearest, // 2. violate it affects replica safety.
// 3-4 affects operator execution speed.
b.preferUpStoreAsLeader, // 3. compare to 4, it is more likely to affect execution speed.
b.preferOldPeerAsLeader, // 4. violate it may or may not affect execution speed.
// 5-7 are less important as they are only trying to build the
// operator with less leader transfer steps.
b.preferAddOrPromoteTargetLeader, // 5. it is precondition of 6 so goes first.
b.preferTargetLeader,             // 6. it may help 7 in later steps.
b.preferLessLeaderTransfer,       // 7. trival optimization to make the operator more tidy.
```

Reply:
update: `avoidRemoveLeader` is removed from the checklist. It is checked in `planLeaders` stage, i.e. not to create plans that removes a leader directly since it is invalid.